oauth2_http/java/com/google/auth/oauth2/IdToken.java

```diff
@@ -0,0 +1,88 @@
+/*
+ * Copyright 2019, Google LLC
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are
+ * met:
+ *
+ *    * Redistributions of source code must retain the above copyright
+ * notice, this list of conditions and the following disclaimer.
+ *    * Redistributions in binary form must reproduce the above
+ * copyright notice, this list of conditions and the following disclaimer
+ * in the documentation and/or other materials provided with the
+ * distribution.
+ *
+ *    * Neither the name of Google LLC. nor the names of its
+ * contributors may be used to endorse or promote products derived from
+ * this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+package com.google.auth.oauth2;
+
+import java.io.Serializable;
+import java.util.Date;
+import java.util.Objects;
+
+import com.google.api.client.json.webtoken.JsonWebSignature;
+import com.google.common.base.MoreObjects;
+
+/**
+ * Represents a temporary IdToken and its JSONWebSingature object
+ */
+public class IdToken extends AccessToken implements Serializable {
+
+  private static final long serialVersionUID = -8514239465808977353L;
+
+  private final JsonWebSignature jws;
+
+  /**
+   * @param tokenValue     String representation of the Id token.
+   * @param jws            JsonWebSignature as object
+   */
+  public IdToken(String tokenValue, JsonWebSignature jws) {
+    super(tokenValue, new Date(jws.getPayload().getExpirationTimeSeconds() * 1000));
+    this.jws = jws;
+  }
+
+  /**
+   * The JsonWebSignature as object
+   *
+   * @return returns com.google.api.client.json.webtoken.JsonWebSignature
+   */
+  public JsonWebSignature getJsonWebSignature() {
+    return jws;
+  }
+
+  @Override
+  public int hashCode() {
+    return Objects.hash(super.getTokenValue(), jws);
```

Comment:
nit: In theory the raw token and the parsed JsonWebSignature should be referencing the same data.

Reply:
yes, it should be but since i'm passing them in separately they maybe off.  i tried putting this check into the constructor just now

```java
  public IdToken(String tokenValue, JsonWebSignature jws) throws IOException{
    super(tokenValue, new Date(jws.getPayload().getExpirationTimeSeconds() * 1000));
    if (!tokenValue.equals(jws.toString())) {
      throw new IOException("Provided IdToken value does not match provided JsonWebSignature");
    }
    this.jws = jws;
  }
```

but `JsonWebSignature.toString()` returns a formatted json string and it doesn't expose an automatic way to return the encoded JWT (hint; it'd be a good api to have in `JsonWebSignature`...

Reply:
It would be easier to make this change in google3's monorepo. Nonetheless, do not rule out adding such a method to JsonWebSignature in google-http-java-client first and then using it here.

Reply:
ok, I'll leave it as-is; we can always ad din that check for coverage later on

Reply:
ref:
  https://github.com/googleapis/google-http-java-client/pull/757